### PR TITLE
feat: add RotatedSurfaceCode and RotatedSurfaceCodeLoom to simulator utilities

### LINKED ISSUE #38: Integrate surface code.
**Problem**
- The staging_surface_code branch (origin/staging_surface_code) adds SurfaceCode and RotatedSurfaceCode classes, but was developed in parallel with PR33. It needs to be rebased onto PR33, and the conditional gate support from t[his issue](https://github.com/CERN-IT-INNOVATION/quantum-gates/issues/37) is needed for the syndrome-correction step.

**Relevant files**:
All on branch staging_surface_code:
- src/quantum_gates/_utility/surface_code.py
- src/quantum_gates/_utility/rotated_surface_code.py
start up and testing/AER_testing_surface_code_SV.ipynb, this is a reference for the expected behaviour

**Solution**
- Create a new branch from main after PR33 and reapply the changes from staging_surface_code.*
- Confirm the surface code circuits run end-to-end with BinaryCircuit (the only circuit class that supports non-linear topologies today)
- Add unit tests for a distance-2 rotated surface code in zero noise: all syndromes should be 0

*Note that this might be much easier than rebasing as we have a large number of commits and some merge conflicts that are easier solved by reapplying the changes on a case by case basis by hand or with AI.

### PR DESCRIPTION
**Pull request by Cherie000**
## Summary

This PR adds two new files to `src/quantum_gates/_utility/` implementing
end-to-end rotated surface code simulations using the MrAnderson simulator.

### `RotatedSurfaceCodeLoom.py`
A surface code simulation pipeline built on top of the `loom` framework.
Uses `EkaToQasmConverter` to generate a Qiskit circuit from a declarative
surface code description, runs it through MrAnderson, and decodes the logical error rate via PyMatching,
using Stim to construct the detector error model.

**Key features:**
- Supports distance-$d$ rotated surface code with configurable cycles, shots, and noise
- Two simulation backends: `MrAnderson` (custom statevector) and `AER` (Qiskit stabilizer)
- Stim-based noise model and detector error model for PyMatching decoding
- Returns logical error rate directly

### `RotatedSurfaceCode.py`
A self-contained rotated surface code implementation built directly in Qiskit,
without the loom framework.

**Key features:**
- Constructs the full surface code lattice geometry (data qubits, X/Z stabilizers, neighbors)
- Builds stabilizer measurement cycles with reset, entanglement, and measurement layers
- Supports optional fault injection via `error` parameter
- Includes a full decoding pipeline: parity-check matrix construction, PyMatching
  decoding per cycle, and Z-correction of data qubits
- Compatible with both MrAnderson (`CustomNoiseChannelsGates`) and Aer backends

## Dependencies added (`docs/requirements.txt`)
- `pymatching`
- `stim`
- `el-loom`

## Notes
- These files serve as concrete end-to-end usage examples of the MrAnderson
  simulator on a fault-tolerant quantum error correction protocol
- `RotatedSurfaceCodeLoom.py` depends on the `loom` ecosystem;
  `RotatedSurfaceCode.py` is standalone
- Both files are intended to live in `src/quantum_gates/_utility/`

### CODE REVIEW

File: docs/requirements.txt
Comment (romanwixinger) [inline docs/requirements.txt:1]: Todo: Check if we still need this full list after the changes I did to the docs in https://github.com/CERN-QTI/quantum-gates/pull/55/changes. I guess the dependencies for the build process of the docs should now come from the setup.cfg.

File: src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
+
```
Comment (romanwixinger) [inline src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py:1]: Todo: Cleanup the formatting of the imports. No extra lines on top and two lines between the last import and the class definition.

File: src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        dem = noisy_circuit.detector_error_model(
+            decompose_errors=True,
+            approximate_disjoint_errors=True,
+        )
+
+        matcher = pymatching.Matching(dem)
+        return matcher, converter
+
+    def MrAnderson_run_circ_backend(self):
```
Comment (romanwixinger) [inline src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py]: Todo: Make method names lowercase. I know that we have the gates uppercase but this is just for legacy reasons and because it is easier to read.

File: src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+class RotatedSurfaceCodeLoom:
+    def __init__(self, distance: int = 3, n_cycles: int = 5, n_shots: int = 1000, noise = False, p = 0.01):
+        self.distance = distance
+        self.n_cycles = n_cycles
+        self.n_shots = n_shots
+        self.noise = noise
+        self.p = p
+
+        # ── Build experiment and interpret ───────────────────────────────────
```
Comment (romanwixinger) [inline src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py:34]: Todo: Cleanup comments,  no extra "─".

File: src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py
Comment (romanwixinger) [inline src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py:1]: Todo: Name the file in lowercase.

File: tests/simulation/test_rotated_surface_code.py
Comment (romanwixinger) [inline tests/simulation/test_rotated_surface_code.py:1]: Question: Do we have any other properties that we could test to verify that it works. For example, are there some trivial cases (circuits with one or a few qubits) in which we know the number of gates that are introduced and could test for it? In my view, in the zero noise regime the surface code should not change the output distribution, so we could test against circuits with/without the surface code right?

File: src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+            coupling_map=cm,
+            initial_layout=list(range(self.n_qubits)),
+            seed_transpiler=42,
+            scheduling_method=needs_controlflow,
+        )
+        return t_circ
+
+    def MrAnderson_run_circ(self):
+        print("n_qubits:", self.n_qubits)
+        backend = FakeBrisbane()
```
Comment (romanwixinger) [inline src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py:190]: Input: Creating the backend here feels a little bit wrong. Later we use it for self._get_device_parameters(backend) and so the question is what kind of information we are adding/implying here? Shouldn't this information come from the outside to the class instance?

File: src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        print("Simulation complete")
+        return res["mid_counts"]
+
+    def AER_run_circ(self):
+        if self.noise:
+            # Build noise model
+            noise_model = NoiseModel()
+
+            # Single-qubit gate depolarizing noise
+            #single_qubit_error = depolarizing_error(p, 1)
```
Comment (romanwixinger) [inline src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py:225]: Todo: Check if still needed.

File: src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                        break
+                
+                # Fall back to average if no nearby edge found
+                if found_p_int is None:
+                    found_p_int = avg_p_int
+                
+                device_param_lookup['p_int'][i, i+1] = found_p_int
+                device_param_lookup['p_int'][i+1, i] = found_p_int
+        return device_param_lookup  
+    def _transpile_circ(self):
```
Comment (romanwixinger) [inline src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py]: Todo: Format the methods, one line between methods.

File: src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+import qiskit.qasm3 as qasm3
+from qiskit import transpile
+from qiskit.circuit.controlflow import ControlFlowOp 
+from qiskit.transpiler import CouplingMap
+from qiskit_ibm_runtime.fake_provider import FakeBrisbane
+from qiskit_aer import AerSimulator
+from qiskit_aer.noise import NoiseModel, depolarizing_error
+
+from loom.eka import Eka, Lattice
```
Comment (romanwixinger) [inline src/quantum_gates/_utility/RotatedSurfaceCodeLoom.py:14]: I like how you organized the imports!

File: tests/simulation/test_rotated_surface_code.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# tests/test_rotated_surface_code.py
+from quantum_gates._utility.RotatedSurfaceCodeLoom import RotatedSurfaceCodeLoom
```
Comment (romanwixinger) [inline tests/simulation/test_rotated_surface_code.py]: Todo: Make sure to expose the new functionality to the outside. The imports in the future should not have to use ._utility "private underscore" for functionality that should be public.

File: src/surface_code_ler.png
Comment (romanwixinger) [inline src/surface_code_ler.png:1]: Todo: Move to the docs.

**Review by romanwixinger — CHANGES_REQUESTED** [review]
Great work @Cherie000 ! Let us discuss how we can test the surface code. Maybe you have ideas from your own experiments that we can tap into.

**Comment by NathanPaceydev** [comment]
Since the SC both rotated and planar implement "large" circuits it would be good to also run a profiler on the classes and see what speed-ups we could apply!
